### PR TITLE
Add optimization to randomize null key in outer join

### PR DESCRIPTION
**Pull request by feilong-liu**
### What's added?

This PR adds an optimization for outer join which has many NULL in join keys. 

For each join key equal pair, `leftCol = rightCol`, it adds a `coalesce(cast(leftCol as varchar), 'l' || cast(random(hashpartitions) as varchar)) = coalesce(cast(rightCol as varchar), 'r' || cast(random(hashpartitions) as varchar))`, so as to avoid skew in NULL values.

### Test plan - (Please fill in how you tested your changes)
Add unit tests

**The failure in native test is not related to this PR**

```
== RELEASE NOTES ==

General Changes
* Add optimization for outer join.
  Add randomized value for NULL join keys to avoid skew in NULL. Default to false.

```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RandomizeNullKeyInOuterJoin.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                        joinClauses,
+                        joinNode.getOutputVariables(),
+                        joinNode.getFilter(),
+                        joinNode.getLeftHashVariable(),
+                        joinNode.getRightHashVariable(),
+                        joinNode.getDistributionType(),
+                        joinNode.getDynamicFilters()));
+    }
+
+    private RowExpression randomizeJoinKey(RowExpression keyExpression, Context context, String prefix)
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RandomizeNullKeyInOuterJoin.java:134]: Move this up next to the other private helper methods

**Review by kaikalur — APPROVED** [review]
Please add a SqlBenchMark that shows both improvement when the issue exists and also any overhead it would cause when there no null keys.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+        private Rewriter(Session session,
+                FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator planNodeIdAllocator, PlanVariableAllocator planVariableAllocator,
+                Map<VariableReferenceExpression, VariableReferenceExpression> keyToRandomKeyMap)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.planNodeIdAllocator = requireNonNull(planNodeIdAllocator, "planNodeIdAllocator is null");
+            this.planVariableAllocator = requireNonNull(planVariableAllocator, "planVariableAllocator is null");
+            this.keyToRandomKeyMap = requireNonNull(keyToRandomKeyMap, "keyToRandomKeyMap is null");
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:115]: Why is this passed in? I thought we use only a single instance for all? I say just initialize it in the class. No need to pass it in. If you do need to pass it in, put it in context.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                    "concat",
+                    VARCHAR,
+                    ImmutableList.of(constant(Slices.utf8Slice(prefix), VARCHAR), randomNumberVarchar));
+
+            RowExpression castToVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, keyExpression.getType(), VARCHAR), VARCHAR, keyExpression);
+            return new SpecialFormExpression(SpecialFormExpression.Form.COALESCE, VARCHAR, ImmutableList.of(castToVarchar, concatExpression));
+        }
+
+        // Do not need to generate randomized variable if the joinKey 1) has already been randomized 2) included in the output of the source node
+        private boolean randomKeyAvailable(PlanNode source, VariableReferenceExpression joinKey)
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:142]: Name it something like: isAlreadyRandomized

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 4]
```diff
@@ -6197,4 +6198,15 @@ public void testOrderByCompilation()
...
+    @Test
+    public void testRandomizeNullKeyOuterJoin()
+    {
+        Session enableRandomize = Session.builder(getSession())
+                .setSystemProperty(RANDOMIZE_OUTER_JOIN_NULL_KEY, "true")
+                .build();
+
+        String leftJoin = "SELECT o.orderkey, l.discount FROM orders o LEFT JOIN lineitem l ON o.orderkey = l.orderkey";
+        assertQuery(enableRandomize, leftJoin, getSession(), leftJoin);
+    }
```
Comment (kaikalur) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:6211]: Also add multiple joins here

**Review by kaikalur — CHANGES_REQUESTED** [review]

**Comment by feilong-liu** [comment]
> Please add a SqlBenchMark that shows both improvement when the issue exists and also any overhead it would cause when there are no null keys.

@kaikalur 
As we discussed offline, there is no existing support for benchmark which runs with distributed query runner and demonstrate performance improvement from optimizations for data distribution.

So I manually generate data sets with different number of null keys and run test queries to compare the performance. The optimization indeed reduce latency when having many null keys but also cost more CPU resources.
Test vs. Control:
When there is no skew: 4.4% more CPU and 4.1% more latency
20% of keys are null: 15% more CPU and 9.8% less in latency
90% of keys are null: 86.2% more CPU and 50.8% less in latency

Adding a comprehensive benchmark take much effort and is out of scope of this PR, and adding a simplified one does not help much here imo, considering that this will be a default to false optimization rule, can we go with the ad-hoc numbers above for now?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RandomizeNullKeyInOuterJoin.java (hunk 1 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RandomizeNullKeyInOuterJoin.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RandomizeNullKeyInOuterJoin.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        if (candidateEquiJoinClauses.isEmpty()) {
+            return Result.empty();
+        }
+
+        List<RowExpression> leftRandomExpressions = candidateEquiJoinClauses.stream().map(x -> x.getLeft()).map(x -> randomizeJoinKey(x, context, "l")).collect(Collectors.toList());
+        List<VariableReferenceExpression> leftRandomVariable = leftRandomExpressions.stream().map(x -> context.getVariableAllocator().newVariable(x, VARIABLE_SUFFIX)).collect(Collectors.toList());
+
+        List<RowExpression> rightRandomExpressions = candidateEquiJoinClauses.stream().map(x -> x.getRight()).map(x -> randomizeJoinKey(x, context, "r")).collect(Collectors.toList());
+        List<VariableReferenceExpression> rightRandomVariable = rightRandomExpressions.stream().map(x -> context.getVariableAllocator().newVariable(x, VARIABLE_SUFFIX)).collect(Collectors.toList());
+
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RandomizeNullKeyInOuterJoin.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RandomizeNullKeyInOuterJoin.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+public class RandomizeNullKeyInOuterJoin
+        implements Rule<JoinNode>
+{
+    private static final String VARIABLE_SUFFIX = RandomizeNullKeyInOuterJoin.class.getSimpleName();
+    private final FunctionAndTypeManager functionAndTypeManager;
+    // If variable name contains VARIABLE_SUFFIX, the join has already gone through optimization. Check to break endless optimization.
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RandomizeNullKeyInOuterJoin.java]: Good suggestion! Updated.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+            }
+
+            List<JoinNode.EquiJoinClause> candidateEquiJoinClauses = joinNode.getCriteria().stream().filter(x -> isSupportedType(x.getLeft()) && isSupportedType(x.getRight())).collect(Collectors.toList());
+            if (candidateEquiJoinClauses.isEmpty()) {
+                return context.defaultRewrite(joinNode, context.get());
+            }
+
+            PlanNode rewrittenLeft = context.rewrite(joinNode.getLeft());
+            PlanNode rewrittenRight = context.rewrite(joinNode.getRight());
+
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:192]: Added the check.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java (hunk 2 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java]: Update the code to initialize the map in class constructor.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java (hunk 3 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java]: Done

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 4 above)
Comment (feilong-liu) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:6214]: Done

**Review by kaikalur — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+            this.planVariableAllocator = requireNonNull(planVariableAllocator, "planVariableAllocator is null");
+            this.keyToRandomKeyMap = new HashMap<>();
+        }
+
+        private static boolean isSupportedType(VariableReferenceExpression variable)
+        {
+            return Stream.of(VARCHAR, BIGINT, DATE).anyMatch(x -> x.equals(variable.getType()));
+        }
+
+        private RowExpression randomizeJoinKey(RowExpression keyExpression, String prefix)
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:103]: private functions go at end of class definition(after public)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+                    .collect(ImmutableMap.toImmutableMap(i -> randomVariable.get(i), i -> randomExpressions.get(i)));
+
+            return result;
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode joinNode, RewriteContext<Void> context)
+        {
+            // For cross join, it does not allow mismatch of input and output. Need to reassign output in case input of the cross join changes.
+            if (joinNode.isCrossJoin()) {
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:147]: Can you explain more on why is this needed(what causes the mismatch)? Should we add a project on top of our rewrite, so we don't have to handle this?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java (hunk 8 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java (hunk 9 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:109]: > Can you explain more on why is this needed(what causes the mismatch)?

This is because we also add the random key to the output of the optimized join node, so that it can be reused if in later stage we do outer join on the same join key.

The cross join's input can be changed if one of its input contains the outer join which are optimized by this rule.

> This seems like it could break in other places too.

Should not. In this rewrite, we are adding to the output, and it will be pruned if not used by later stages.

> Should we add a project on top of our rewrite, so we don't have to handle this?

If we add a project on top the rewrite to project off the random key added, we will not be able to reuse the random key at all.

Another way to do it is to add project node between the cross join and its source nodes, and project off the random key. This way the random key will not be available after the cross join and not able to reuse after it.

**Comment by kaikalur** [comment]
> > Please add a SqlBenchMark that shows both improvement when the issue exists and also any overhead it would cause when there are no null keys.
> 
> @kaikalur As we discussed offline, there is no existing support for benchmark which runs with distributed query runner and demonstrate performance improvement from optimizations for data distribution.
> 
> So I manually generate data sets with different number of null keys and run test queries to compare the performance. The optimization indeed reduce latency when having many null keys but also cost more CPU resources. Test vs. Control: When there is no skew: 4.4% more CPU and 4.1% more latency 20% of keys are null: 15% more CPU and 9.8% less in latency 90% of keys are null: 86.2% more CPU and 50.8% less in latency
> 
> Adding a comprehensive benchmark take much effort and is out of scope of this PR, and adding a simplified one does not help much here imo, considering that this will be a default to false optimization rule, can we go with the ad-hoc numbers above for now?

86% more cpu!! We should take a look

**Comment by kaikalur** [comment]
Also, looking closely - you don't need to keep the original keys as join keys anymore! So just drop them

**Comment by feilong-liu** [comment]
> Also, looking closely - you don't need to keep the original keys as join keys anymore! So just drop them

Done. Building and running benchmark queries and will report back.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java (hunk 9 above)
Comment (pranjalssh) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:109]: Thanks for the detailed explanation, lets add a test for this.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+                checkState(inputVariables.containsAll(joinNode.getOutputVariables()));
+                return new JoinNode(
+                        joinNode.getSourceLocation(),
+                        joinNode.getId(),
+                        joinNode.getStatsEquivalentPlanNode(),
+                        joinNode.getType(),
+                        rewrittenLeft,
+                        rewrittenRight,
+                        joinNode.getCriteria(),
+                        inputVariables.stream().collect(Collectors.toList()),
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:125]: Use `toImmutableList()` here and other places below

**Comment by feilong-liu** [comment]
> Also, looking closely - you don't need to keep the original keys as join keys anymore! So just drop them

After removing the original join key, for 90% of keys are null, CPU time increases by 28%, latency improves by 69%.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java (hunk 10 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java]: Done

**Comment by feilong-liu** [comment]
> Thanks for the detailed explanation, lets add a test for this.

Tests added for cross join case.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java (hunk 9 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:110]: Tests added for cross join case.

**Review by pranjalssh — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+                    BIGINT,
+                    constant((long) partitionCount, BIGINT));
+            RowExpression randomNumberVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, randomNumber.getType(), VARCHAR), VARCHAR, randomNumber);
+            RowExpression concatExpression = call(functionAndTypeManager,
+                    "concat",
+                    VARCHAR,
+                    ImmutableList.of(constant(Slices.utf8Slice(prefix), VARCHAR), randomNumberVarchar));
+
+            RowExpression castToVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, keyExpression.getType(), VARCHAR), VARCHAR, keyExpression);
+            return new SpecialFormExpression(SpecialFormExpression.Form.COALESCE, VARCHAR, ImmutableList.of(castToVarchar, concatExpression));
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:218]: Let's try to avoid casting if the key is already a varchar

**Review by kaikalur — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+    public RandomizeNullKeyInOuterJoin(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (getJoinDistributionType(session).canPartition() && randomizeOuterJoinNullKeyEnabled(session)) {
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:74]: In this optimization, we add an additional project node in the source of join to generate the randomized variable and add the randomized variable to join output so that it can be reused in later stages. We rely on optimizations in later stages to prune/combine/inline projections to simplify the plan. Move the optimization after DetermineJoinDistributionType can potentially lose some optimizations below. Considering that this optimization is disabled by default and only enabled case by case. I think it may be OK to put it here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+                        joinNode.getCriteria(),
+                        inputVariables.stream().collect(ImmutableList.toImmutableList()),
+                        joinNode.getFilter(),
+                        joinNode.getLeftHashVariable(),
+                        joinNode.getRightHashVariable(),
+                        joinNode.getDistributionType(),
+                        joinNode.getDynamicFilters());
+            }
+
+            if (Stream.of(LEFT, RIGHT, FULL).noneMatch(joinType -> joinType.equals(joinNode.getType()))) {
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:137]: If we move the `if (Stream.of(LEFT, RIGHT, FULL).noneMatch(joinType -> joinType.equals(joinNode.getType()))) ` part to top, we need to add the `!joinNode.isCrossJoin()` to the check condition, which leads to `if (Stream.of(LEFT, RIGHT, FULL).noneMatch(joinType -> joinType.equals(joinNode.getType())) && !joinNode.isCrossJoin())`. Similar for the following case. From code readability and flow of code perspective, I find it seems not helping here due to the additional check of `!isCrossJoin`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+            Map<VariableReferenceExpression, RowExpression> rightKeyRandomVariableMap = generateRandomKeyMap(rightJoinKeys, RIGHT_PREFIX);
+
+            ImmutableList.Builder<JoinNode.EquiJoinClause> joinClauseBuilder = ImmutableList.builder();
+            // Rewrite supported join clauses
+            List<JoinNode.EquiJoinClause> rewrittenJoinClauses = candidateEquiJoinClauses.stream()
+                    .map(x -> new JoinNode.EquiJoinClause(keyToRandomKeyMap.get(LEFT_PREFIX).get(x.getLeft()), keyToRandomKeyMap.get(RIGHT_PREFIX).get(x.getRight())))
+                    .collect(ImmutableList.toImmutableList());
+            joinClauseBuilder.addAll(rewrittenJoinClauses);
+            // Add the join clauses which are not supported back
+            List<JoinNode.EquiJoinClause> unChangedJoinClauses = joinNode.getCriteria().stream()
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+                    .filter(x -> !candidateEquiJoinClauses.contains(x))
+                    .collect(ImmutableList.toImmutableList());
+            joinClauseBuilder.addAll(unChangedJoinClauses);
+
+            Assignments.Builder leftAssignments = Assignments.builder();
+            leftAssignments.putAll(leftKeyRandomVariableMap);
+            leftAssignments.putAll(rewrittenLeft.getOutputVariables().stream().collect(ImmutableMap.toImmutableMap(identity(), identity())));
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+                    "concat",
+                    VARCHAR,
+                    ImmutableList.of(constant(Slices.utf8Slice(prefix), VARCHAR), randomNumberVarchar));
+
+            RowExpression castToVarchar = keyExpression;
+            // Only do cast if keyExpression is not VARCHAR type.
+            if (!(keyExpression.getType() instanceof VarcharType)) {
+                castToVarchar = call("CAST", functionAndTypeManager.lookupCast(CAST, keyExpression.getType(), VARCHAR), VARCHAR, keyExpression);
+            }
+            return new SpecialFormExpression(SpecialFormExpression.Form.COALESCE, VARCHAR, ImmutableList.of(castToVarchar, concatExpression));
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+        private boolean isAlreadyRandomized(PlanNode source, VariableReferenceExpression joinKey, String prefix)
+        {
+            return keyToRandomKeyMap.containsKey(prefix) && keyToRandomKeyMap.get(prefix).containsKey(joinKey) && source.getOutputVariables().contains(keyToRandomKeyMap.get(prefix).get(joinKey));
+        }
+
+        private Map<VariableReferenceExpression, RowExpression> generateRandomKeyMap(List<VariableReferenceExpression> joinKeys, String prefix)
+        {
+            List<RowExpression> randomExpressions = joinKeys.stream().map(x -> randomizeJoinKey(x, prefix)).collect(ImmutableList.toImmutableList());
+            List<VariableReferenceExpression> randomVariable = randomExpressions.stream()
+                    .map(x -> planVariableAllocator.newVariable(x, RandomizeNullKeyInOuterJoin.class.getSimpleName()))
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:238]: Since this is generated by optimization rules, I intentionally name it with this hint so that it can stand out and easy to identify when debugging query plans.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestRandomizeNullKeyInOuterJoin.java [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+                                                ImmutableMap.of("leftCol", expression("leftCol"), "leftRandom", expression("coalesce(cast(leftCol as varchar), 'l' || cast(random(100) as varchar))")),
+                                                tableScan("orders", ImmutableMap.of("leftCol", "orderkey")))),
+                                anyTree(
+                                        project(
+                                                ImmutableMap.of("rightCol", expression("rightCol"), "rightRandom", expression("coalesce(cast(rightCol as varchar), 'r' || cast(random(100) as varchar))")),
+                                                tableScan("lineitem", ImmutableMap.of("rightCol", "orderkey")))))));
+    }
+
+    @Test
+    public void testLeftJoinOnSameKey()
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestRandomizeNullKeyInOuterJoin.java:82]: Tests added.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:74]: we can just add the inline projection rules right after it.  I think it would be better to only do it for partitioned joins.  Especially consider that you can have a query that has a mix of join types.  Also, this is the kind of thing that ideally in the future we'd make a cost-based decision about, so would be good to have it where we have the info we might need.

[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:76]: Updated the code to put the rule after DetermineJoinDistributionType and DetermineSemiJoinDistributionType rules.

**Review by rschlussel — APPROVED** [review]
